File: src/main/java/com/fireforest/state/CellState.java
package com.fireforest.state;

import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import com.fireforest.entity.Cell;

@JsonTypeInfo(
        use = JsonTypeInfo.Id.NAME,
        include = JsonTypeInfo.As.PROPERTY,
        property = "type"
)
@JsonSubTypes({
        @JsonSubTypes.Type(value = FireState.class, name = "fire"),
        @JsonSubTypes.Type(value = TreeState.class, name = "tree"),
        @JsonSubTypes.Type(value = AshState.class, name = "ash")
})
public interface CellState {
    void handle(Cell cell, Cell[][] forest, double probability);
    String getSymbol();
    void tryIgnite(Cell cell, double probability);
}

File: fireforest-client/src/types/CellState.ts
export type CellType = "fire" | "tree" | "ash";

export interface CellState {
  type: CellType;
  symbol: string;
}

export interface Cell {
  x: number;
  y: number;
  currentState: CellState;
}

export interface Grid {
  height: number;
  width: number;
  grid: Cell[][];
}
